mirrord-agent/src/tcp/outgoing.rs

```diff
@@ -0,0 +1,228 @@
+use core::fmt;
+use std::{collections::HashMap, path::PathBuf};
+
+use mirrord_protocol::{
+    tcp::outgoing::{
+        ConnectRequest, ConnectResponse, ReadResponse, TcpOutgoingRequest, TcpOutgoingResponse,
+        WriteRequest, WriteResponse,
+    },
+    RemoteResult,
+};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpStream,
+    select,
+    sync::mpsc::{self, Receiver, Sender},
+    task,
+};
+use tracing::{debug, error, trace, warn};
+
+use crate::{error::AgentError, runtime::set_namespace};
+
+type Request = TcpOutgoingRequest;
+type Response = TcpOutgoingResponse;
+
+pub(crate) struct TcpOutgoingApi {
+    task: task::JoinHandle<Result<(), AgentError>>,
+    request_channel_tx: Sender<Request>,
+    response_channel_rx: Receiver<Response>,
+}
+
+pub struct Data {
+    id: i32,
+    bytes: Vec<u8>,
+}
+
+impl fmt::Debug for Data {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Data")
+            .field("id", &self.id)
+            .field("bytes (length)", &self.bytes.len())
+            .finish()
+    }
+}
+
+impl TcpOutgoingApi {
+    pub(crate) fn new(pid: Option<u64>) -> Self {
+        let (request_channel_tx, request_channel_rx) = mpsc::channel(1000);
+        let (response_channel_tx, response_channel_rx) = mpsc::channel(1000);
+
+        let task = task::spawn(Self::request_task(
+            pid,
+            request_channel_rx,
+            response_channel_tx,
+        ));
+
+        Self {
+            task,
+            request_channel_tx,
+            response_channel_rx,
+        }
+    }
+
+    async fn interceptor_task(
+        connection_id: i32,
+        response_tx: Sender<Response>,
+        mut write_rx: Receiver<Data>,
+        stream: TcpStream,
+    ) {
+        trace!("intercept_task -> id {:#?}", connection_id);
+
+        let mut buffer = vec![0; 1500];
+        let (remote_reader, mut remote_writer) = stream.into_split();
+
+        loop {
+            select! {
+                biased;
+
+                readable = remote_reader.readable() => {
```

Comment:
I think you can use `AsyncReadExt` to avoid the readable, match flow (as it does it for you)